pdns/syncres.cc

```diff
@@ -2699,11 +2700,25 @@ int SyncRes::doResolveAt(NsSet &nameservers, DNSName auth, bool flawedNSSet, con
   return -1;
 }
 
-void SyncRes::setIncomingECS(boost::optional<const EDNSSubnetOpts&> incomingECS)
+void SyncRes::setQuerySource(const ComboAddress& requestor, boost::optional<const EDNSSubnetOpts&> incomingECS)
 {
-  d_incomingECS = incomingECS;
-  if (incomingECS) {
-    if (d_incomingECS->source.getBits() == 0) {
+  d_requestor = requestor;
+
+  if (incomingECS && incomingECS->source.getBits() > 0) {
+    d_cacheRemote = incomingECS->source.getMaskedNetwork();
+    uint8_t bits = std::min(incomingECS->source.getBits(), (incomingECS->source.isIpv4() ? s_ecsipv4limit : s_ecsipv6limit));
+    ComboAddress trunc = incomingECS->source.getNetwork();
+    trunc.truncate(bits);
+    d_outgoingECSNetwork = boost::optional<Netmask>(Netmask(trunc, bits));
+  } else {
+    d_cacheRemote = d_requestor;
+    if(s_ednslocalsubnets.match(d_requestor)) {
```

Comment:
It looks like we don't honor an incoming `ECS` with a 0-mask. If `use-incoming-edns-subnet` is set and we receive a 0-mask, we should use `s_ecsScopeZero` even if the source address is in `s_ednslocalsubnets`.

Reply:
Right. I see (now) the RFC mandates this. OTOH I've got a production report that passing on 0.0.0.0 confuses (some) upstreams.

Reply:
Yes, that's why we try hard not to use that in `s_ecsScopeZero`. Not hard enough, perhaps?